core/src/main/scala/org/apache/spark/deploy/StandaloneResourceUtils.scala

```diff
@@ -0,0 +1,356 @@
+              allocated = valid
+              origAllocation = origAllocation.filter(
+                allocation => !invalid.contains(allocation.pid))
+              checked = true
+              // note this is a meaningless return value, just to avoid creating any new object
+              available
+            } else {
+              available.take(amount)
+            }
+          }
+          rName -> assigned
+        }.toMap
+      }
+      val newAllocation = {
+        val allocations = newAssignments.map { case (rName, addresses) =>
+          ResourceAllocation(ResourceID(componentName, rName), addresses)
+        }.toSeq
+        StandaloneResourceAllocation(pid, allocations)
+      }
+      writeResourceAllocationJson(
+        componentName, origAllocation ++ Seq(newAllocation), resourcesFile)
+      newAllocation.toResourceInformationMap
+    } finally {
+      releaseLock(lock)
+    }
+  }
+
+  /**
+   * Free the indicated resources to make those resources be available for other
+   * workers/drivers on the same host.
+   * @param conf SparkConf
+   * @param componentName spark.driver / spark.worker
+   * @param toRelease the resources expected to release
+   * @param pid the process id of worker/driver to release resources.
+   */
+  def releaseResources(
+      conf: SparkConf,
+      componentName: String,
+      toRelease: Map[String, ResourceInformation],
+      pid: Int)
+    : Unit = {
+    if (toRelease != null && toRelease.nonEmpty) {
+      val lock = acquireLock(conf)
+      try {
+        val resourcesFile = new File(getOrCreateResourcesDir(conf), ALLOCATED_RESOURCES_FILE)
+        if (resourcesFile.exists()) {
+          val (target, others) =
+            allocatedStandaloneResources(resourcesFile.getPath).partition(_.pid == pid)
+          if (target.nonEmpty) {
+            val rNameToAddresses = {
+              target.head.allocations.map { allocation =>
+                allocation.id.resourceName -> allocation.addresses
+              }.toMap
+            }
+            val allocations = {
+              rNameToAddresses.map { case (rName, addresses) =>
+                val retained = addresses.diff(Option(toRelease.getOrElse(rName, null))
+                  .map(_.addresses).getOrElse(Array.empty))
+                rName -> retained
+              }
+                .filter(_._2.nonEmpty)
+                .map { case (rName, addresses) =>
+                  ResourceAllocation(ResourceID(componentName, rName), addresses)
+                }.toSeq
+            }
+            if (allocations.nonEmpty) {
+              val newAllocation = StandaloneResourceAllocation(pid, allocations)
+              writeResourceAllocationJson(
+                componentName, others ++ Seq(newAllocation), resourcesFile)
+            } else {
+              if (others.isEmpty) {
+                if (!resourcesFile.delete()) {
+                  logWarning(s"Failed to delete $ALLOCATED_RESOURCES_FILE.")
```

Comment:
We don't delete the resourcesFile while it still has the allocations. And this is more likely the IO error if fails to delete. And it would be OK if we failed to delete the file because now we use pid to detect not valid allocations.